core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java

```diff
@@ -728,6 +728,7 @@ public void delete(String tableName)
     List<ByteBuffer> args = Arrays.asList(ByteBuffer.wrap(tableName.getBytes(UTF_8)));
     Map<String,String> opts = new HashMap<>();
     try {
+      cancelCompaction(tableName);
```

Comment:
To avoid certain race conditions, it would be nice if we did something like the following to only resolve the tableName to a tableId once.  However I don't see an easy way to do this.

 1. tableId = resolveTableId(tableName)
 2. cancelCompaction(tableId)
 3. delete(tableId)

Looking around in the code, this would be easy for cancelCompaction because it already resolves the table id client side.  So could create `private cancelCompaction(TableId)` used by this method and the public `cancelCompaction(String)` method.   However delete table seems like its currently passing the table name to the server side where it would resolve the table name to a table id server side. I may have missed it but I tried to follow the code to the thrift call and did not see anything on the client side resolving the table id for delete table.

Reply:
If were to change the code pass a tableId to the server instead of a tablename, this could cause thrift incompat problems.

Reply:
I've been wanting to change the client code to always resolve tableId client side and change the RPC to only take tableIds, and never tableName (except as second argument in clone or rename, obviously).

My question about this change is: does cancelCompaction(tableName) wait? Does it time out?

Reply:
> That may be a good follow on issue

I agree. I only wanted to give additional support for the idea to do it here, because it would be consistent with potential future work.

Reply:
> does cancelCompaction(tableName) wait? Does it time out?

cancelCompaction(tableName) will wait if there is another process that holds the CANCEL_COMPACT Fate lock. The Fate object just updates ZK, I think the probability of having to wait is low (how many tables are being deleted concurrently?) and if it did need to wait, I think the time would be small. I don't know if Fate transactions time out.

Reply:
The cancel compaction waits on a table read lock.  The delete operation waits on a table write lock.  So anything that cancel would wait on, delete would also wait on.  Could have longer wait times if another fate op comes in between the cancel and delete.

Reply:
@ctubbsii - any issues here? You asked about waits and timeouts.

Reply:
@dlmarion The main thing I was concerned about calling cancelCompaction here is what happens if cancelCompaction API behaves weirdly. For example, if the cancelCompaction doesn't wait, what happens if the deleteTable FaTE operation gets scheduled first. If it does wait, what happens if the cancelCompaction never returns and the deleteTable FaTE operation never gets submitted?

Also, I was wondering, would it make sense to just do this server-side so there's only one RPC call?

Reply:
I can't speak to the scheduling of FATE operations. The cancelCompaction waits on a table read lock, which it will not get if there is a another process that has the write lock. If the cancelCompaction never returns and the delete never gets submitted I don't think there is any harm here. Ultimately, the user wants to delete the table. I assume the user would realize the table didn't get deleted and try again.

Reply:
> If the cancelCompaction never returns and the delete never gets submitted I don't think there is any harm here.

I disagree with this strongly.

Specifically, I'm concerned about the impact on user experience for injecting another client-server RPC call and series of ZK operations in the pipeline prior to the user's intent to delete the table being recorded as a new FaTE. If anything happens in that pipeline that hangs, the user's intent to delete the table is never recorded... and that could be very frustrating.... like hitting the button on a crosswalk or elevator and nothing happening.

This could be especially frustrating because the user may be deleting as a consequence of prior table failures, and delete is being done as a last resort for recovery (or as the easiest resort, in the case of tables a user considers disposable). Because delete is so destructive, it is often used as a hammer when a table is in such a bad state as to need starting over. Having another operation precede it means that there are additional failure points that can prevent this hammer from being swung.

As for the earlier question:

> how many tables are being deleted concurrently?

I know of some users who create and delete tables at a pace most would consider extreme. I'm not sure how these changes would impact them.

Reply:
Since all we're doing is using the read lock to update a ZK entry, I'm wondering if we can reduce the number of intermediate operations by performing this step early in the delete FaTE RepOs, rather than doing this client side. Can't the delete table RepO get the read lock first, update the ZK entry, then wait on the write lock?

Reply:
That's an interesting thought that I did not explore. I can look at that. I did look at having the Delete fate op call the cancelCompaction fate op, and that seemed overly complex.

Reply:
@ctubbsii  - I backed out my change to TableOperationsImpl and modified the DeleteTable fate transaction in the manner you described.

Reply:
Cool. Thanks for indulging my concerns. Although they were probably not too important, I do think your current implementation is better.